fix: time selector separator in timepicker is not centered vertically

pranavo72bex opened the pull request:
PR Description
This PR addresses an issue in the Flutter time picker where the separator between the hour and minute selectors (:) was not centered vertically. The change ensures that the separator is now vertically centered, improving the overall UI alignment and consistency.

Previously, the separator used only the TextAlign.center property, which did not fully address the vertical alignment issue. The updated code wraps the separator text in a Center widget to ensure proper centering both horizontally and vertically.

Before:

The separator (:) was horizontally centered but not aligned vertically.
After:

The separator is now both horizontally and vertically centered within the time picker.

![image-20241023-041530](https://github.com/user-attachments/assets/8401264a-a751-47b4-b634-5742ba4fabcd) ![Screenshot 2024-10-23 at 12 33 05](https://github.com/user-attachments/assets/ab95abd9-8de8-4f7d-9114-d96073d8f887)


Pre-launch Checklist
 I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
 I read the [Tree Hygiene] wiki page, which explains my responsibilities.
 I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
 I signed the [CLA].
 I listed at least one issue that this PR fixes in the description above.
 I updated/added relevant documentation (doc comments with ///).
 I added new tests to check the change I am making, or this PR is [test-exempt].
 I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
 All existing and new tests are passing.
For further assistance, you can reach out on the #hackers-new channel on [Discord].

pranavo72bex commented:
> @pranavo72bex thank you for putting together this PR. What issue is this fixing? The linked issue is the placeholder, and I couldn't find the issue by searching.
> 
> Also this will need a regression test to ensure your fix isn't overwritten in the future. If there is a simple reproduction of the issue, then we can just test that and see if it's centered. If that's not possible, at worst we can add a golden file test since it looks like the time picker doesn't have one yet. It's fairly straight forward to do so, though kind of odd if you aren't used to it. [This PR adds a similar golden test for a different widget](https://github.com/flutter/flutter/pull/139236), though ignore the leak tracking and scale.
> 
> More in depth on goldens here: https://github.com/flutter/flutter/blob/master/docs/contributing/testing/Writing-a-golden-file-test-for-package-flutter.md#creating-a-new-golden-file-test


Thank you for reviewing the PR !. Here is an example code snippet and a preview to help illustrate the fix 
```
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Time Picker Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        useMaterial3: false,
      ),
      home: TimePickerExample(),
    );
  }
}

class TimePickerExample extends StatefulWidget {
  @override
  _TimePickerExampleState createState() => _TimePickerExampleState();
}

class _TimePickerExampleState extends State<TimePickerExample> {
  TimeOfDay? _selectedTime;

  Future<void> _selectTime(BuildContext context) async {
    final TimeOfDay? pickedTime = await showTimePicker(
      context: context,
      initialTime: _selectedTime ?? TimeOfDay.now(),
    );
    if (pickedTime != null && pickedTime != _selectedTime) {
      setState(() {
        _selectedTime = pickedTime;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Time Picker Example'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              _selectedTime != null
                  ? 'Selected time: ${_selectedTime!.format(context)}'
                  : 'No time selected',
              style: TextStyle(fontSize: 18),
            ),
            SizedBox(height: 20),
            ElevatedButton(
              onPressed: () => _selectTime(context),
              child: Text('Select Time'),
            ),
          ],
        ),
      ),
    );
  }
}
```

![Screenshot 2024-10-31 at 18 57 35](https://github.com/user-attachments/assets/4b8d92e4-789e-4424-a80c-225011fa7281)

MitchellGoodwin commented:
Thank you for the reproduction. I think for this PR, it'd be best to add a golden test as mentioned in my last comment. Can you do so?

pranavo72bex commented:
> Thank you for the reproduction. I think for this PR, it'd be best to add a golden test as mentioned in my last comment. Can you do so?

Thank you for the feedback! I've added a golden test as requested to verify the corrected alignment of the separator (:) in the TimePickerDialog.

The test you provided is structured as follows:

```
testWidgets('TimePicker dialog displays centered separator between hour and minute selectors', (WidgetTester tester) async {
    tester.view.physicalSize = const Size(400, 800);
    tester.view.devicePixelRatio = 1.0;

    await tester.pumpWidget(
      MaterialApp(
        theme: ThemeData(useMaterial3: false),
        home: const MediaQuery(
          data: MediaQueryData(),
          child: Material(
            child: TimePickerDialog(
              initialTime: TimeOfDay(hour: 12, minute: 0),
            ),
          ),
        ),
      ),
    );
    await tester.pumpAndSettle();

    await expectLater(
      find.byType(Dialog),
      matchesGoldenFile('time_picker.dialog.separator.alignment.corrected.png'),
    );
  });
```
This test captures the TimePickerDialog in Material 2 mode (useMaterial3: false) to verify that the separator (:) between the hour and minute selectors is now centered both horizontally and vertically, resolving the previous alignment issue. The generated golden file (time_picker.dialog.separator.alignment.corrected.png) reflects this corrected positioning for consistent UI verification.

Here's a screenshot:
![material time_picker dialog separator alignment corrected](https://github.com/user-attachments/assets/7cc1c2f6-0320-4bd1-ad00-cdfbaba99d9a)

MitchellGoodwin commented:
Analyzer is complaining about needing to be marked as a part of the reduced test set. Please add to the top of the file, above the library and `@TestOn('!chrome')` the tag for it. So it should look like this:

```
// This file is run as part of a reduced test set in CI on Mac and Windows
// machines.
@Tags(<String>['reduced-test-set'])
@TestOn('!chrome')
library;
```

In the Material tests, the file floating_action_button_test.dart is setup the same way.

MitchellGoodwin commented:
> This test captures the TimePickerDialog in Material 2 mode (useMaterial3: false) to verify that the separator (:) between the hour and minute selectors is now centered both horizontally and vertically, resolving the previous alignment issue.

Is this not an issue in Material 3?

pranavo72bex commented:
> Analyzer is complaining about needing to be marked as a part of the reduced test set. Please add to the top of the file, above the library and `@TestOn('!chrome')` the tag for it. So it should look like this:
> 
> ```
> // This file is run as part of a reduced test set in CI on Mac and Windows
> // machines.
> @Tags(<String>['reduced-test-set'])
> @TestOn('!chrome')
> library;
> ```
> 
> In the Material tests, the file floating_action_button_test.dart is setup the same way.

Thank you for you feedback. Please verify the change 🙏

pranavo72bex commented:
> > This test captures the TimePickerDialog in Material 2 mode (useMaterial3: false) to verify that the separator (:) between the hour and minute selectors is now centered both horizontally and vertically, resolving the previous alignment issue.
> 
> Is this not an issue in Material 3?

It is also affecting Material 3 too as shown in screenshot below 
![Screenshot 2024-11-05 at 09 29 35](https://github.com/user-attachments/assets/6ec8a15d-810a-4005-b5e4-c73790b78153)

Code to reproduce issue 
```
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Time Picker Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        // useMaterial3: false,
      ),
      home: TimePickerExample(),
    );
  }
}

class TimePickerExample extends StatefulWidget {
  @override
  _TimePickerExampleState createState() => _TimePickerExampleState();
}

class _TimePickerExampleState extends State<TimePickerExample> {
  TimeOfDay? _selectedTime;

  Future<void> _selectTime(BuildContext context) async {
    final TimeOfDay? pickedTime = await showTimePicker(
      context: context,
      initialTime: _selectedTime ?? TimeOfDay.now(),
    );
    if (pickedTime != null && pickedTime != _selectedTime) {
      setState(() {
        _selectedTime = pickedTime;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Time Picker Example'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              _selectedTime != null
                  ? 'Selected time: ${_selectedTime!.format(context)}'
                  : 'No time selected',
              style: TextStyle(fontSize: 18),
            ),
            SizedBox(height: 20),
            ElevatedButton(
              onPressed: () => _selectTime(context),
              child: Text('Select Time'),
            ),
          ],
        ),
      ),
    );
  }
}
```

MitchellGoodwin reviewed (commented):
Sorry I let this sit for a bit, it slipped out of my review queue.

I have a small nit on golden file naming, but other than that, this is good to go.

> It is also affecting Material 3 too as shown in screenshot below

If it's affecting both, for completion you could add a Material 3 test as well easily by just copying your test, turning on the `useMaterial3` and naming the other golden test the same as the other, but starting with `m3`.

pranavo72bex commented on packages/flutter/test/material/time_picker_test.dart:
```diff
@@ -1341,6 +1344,31 @@ void main() {
+
+        await expectLater(
+          find.byType(Dialog),
+          matchesGoldenFile('time_picker.dialog.separator.alignment.corrected.png'),
```
Thank you for you feedback. Please verify the change 🙏

MitchellGoodwin commented on packages/flutter/test/material/time_picker_test.dart:
```diff
@@ -1341,6 +1344,55 @@ void main() {
+          find.byType(Dialog),
+          matchesGoldenFile('m2_time_picker.dialog.separator.alignment.png'),
+        );
+      });
+    testWidgets('TimePicker dialog displays centered separator between hour and minute selectors', (WidgetTester tester) async {
```
Nit: please add a line between the two tests. Also, it looks like the indentation of the first line of the new test is off.

Other than that, looks good.

pranavo72bex replied:
Thank you for your feedback. Could you kindly review the change 🙏

MitchellGoodwin commented on packages/flutter/test/material/time_picker_test.dart:
```diff
@@ -1341,6 +1344,56 @@ void main() {
+      });
+
+      testWidgets('TimePicker dialog displays centered separator between hour and minute selectors', (WidgetTester tester) async {
+            tester.view.physicalSize = const Size(400, 800);
```
Nit: body of this test is indented a little too much.

pranavo72bex replied:
Thank you for your feedback. @Piinks  has resolved the indentation issue. Thanks you @Piinks  😄

pranavo72bex replied:
I attempted to rebase the branch again, but the checks are still failing. Would you be able to take over this PR and resolve the issue? Please let me know if you need any context or assistance from my side.

Thank you for your help 🙏

MitchellGoodwin replied:
> I attempted to rebase the branch again, but the checks are still failing. Would you be able to take over this PR and resolve the issue? Please let me know if you need any context or assistance from my side.
> 
> Thank you for your help 🙏

I think the last comment I left should fix the checks, then we can still merge it under your name. If that doesn't work and you want somebody to take over after that, we can still do that.

MitchellGoodwin reviewed (commented):
Thank you for the update. Looks like the test formatting is still a little off. Also, could you please rebase this PR? The analyzer is complaining.

Piinks commented:
Rebased to fix some of the failing checks. 👍

MitchellGoodwin commented on packages/flutter/test/material/time_picker_test.dart:
```diff
@@ -1341,6 +1344,56 @@ void main() {
+          matchesGoldenFile('m2_time_picker.dialog.separator.alignment.png'),
+        );
+      });
+
+      testWidgets('TimePicker dialog displays centered separator between hour and minute selectors', (WidgetTester tester) async {
+            tester.view.physicalSize = const Size(400, 800);
+            tester.view.devicePixelRatio = 1.0;
+
+            await tester.pumpWidget(
+              MaterialApp(
+                theme: ThemeData(useMaterial3: true),
+                home: const MediaQuery(
+                  data: MediaQueryData(),
+                  child: Material(
+                    child: TimePickerDialog(
+                      initialTime: TimeOfDay(hour: 12, minute: 0),
+                    ),
+                  ),
+                ),
+              ),
+            );
+            await tester.pumpAndSettle();
+
+            await expectLater(
+              find.byType(Dialog),
+              matchesGoldenFile('m3_time_picker.dialog.separator.alignment.png'),
+            );
+      });
+
```
```suggestion

      testWidgets('TimePicker dialog displays centered separator between hour and minute selectors', (WidgetTester tester) async {
        tester.view.physicalSize = const Size(400, 800);
        tester.view.devicePixelRatio = 1.0;

        await tester.pumpWidget(
          MaterialApp(
            theme: ThemeData(useMaterial3: true),
            home: const MediaQuery(
              data: MediaQueryData(),
              child: Material(
                child: TimePickerDialog(
                  initialTime: TimeOfDay(hour: 12, minute: 0),
                ),
              ),
            ),
          ),
        );
        await tester.pumpAndSettle();

        await expectLater(
          find.byType(Dialog),
          matchesGoldenFile('m3_time_picker.dialog.separator.alignment.png'),
        );
      });

```

MitchellGoodwin replied:
Indentation is still off a bit. If accepted, may want to check locally that this looks correct. It is difficult to format on GitHub.

MitchellGoodwin commented on packages/flutter/test/material/time_picker_test.dart:
```diff
@@ -1341,6 +1344,56 @@ void main() {
       });
 
+      testWidgets('TimePicker dialog displays centered separator between hour and minute selectors', (WidgetTester tester) async {
+        tester.view.physicalSize = const Size(400, 800);
+        tester.view.devicePixelRatio = 1.0;
+
```
```suggestion
        tester.view.physicalSize = const Size(400, 800);
        tester.view.devicePixelRatio = 1.0;
        addTearDown(tester.view.reset);

```

MitchellGoodwin commented on packages/flutter/test/material/time_picker_test.dart:
```diff
@@ -1341,6 +1344,56 @@ void main() {
+      });
+
+      testWidgets('TimePicker dialog displays centered separator between hour and minute selectors', (WidgetTester tester) async {
+        tester.view.physicalSize = const Size(400, 800);
+        tester.view.devicePixelRatio = 1.0;
+
```
```suggestion
        tester.view.physicalSize = const Size(400, 800);
        tester.view.devicePixelRatio = 1.0;
        addTearDown(tester.view.reset);

```

MitchellGoodwin reviewed (commented):
Did some digging. The problem was that the resized testing view was persisting sometimes to other tests, and [this test](https://github.com/flutter/flutter/blob/95fa7734b3d1701cdd16be3685b9e8f7e3596f15/packages/flutter/test/material/time_picker_test.dart#L633) is expecting the view to be a certain width. So adding a `addTearDown(tester.view.reset)` to both tests *should* fix the checks. Then this will be good to go.

MitchellGoodwin reviewed (approved):
LGTM! Thank you for being so responsive to my nits. This will need a second approval. I'll ask around to see about getting another reviewer.

dkwingsmt reviewed (approved):
LGTM

MitchellGoodwin commented:
> Should the separator be baseline-aligned instead of center-aligned, since it's text?

@QuncCccccc and I were looking into that with the Material specs, especially for things like localization, but it seems like center aligning fits the specs better.

LongCatIsLooong commented:
> > Should the separator be baseline-aligned instead of center-aligned, since it's text?
> 
> @QuncCccccc and I were looking into that with the Material specs, especially for things like localization, but it seems like center aligning fits the specs better.

Ah thanks for looking into that. How does it look in a localized app? There are localizations that don't use `:` as the separator I would assume?

MitchellGoodwin commented:
> Ah thanks for looking into that. How does it look in a localized app? There are localizations that don't use : as the separator I would assume?

Yes, some use '.', others use a 'h'. The dot was the one I was concerned about.

MitchellGoodwin commented:
Google testing failures are for scuba testing, which was expected. Looking into them.

pranavo72bex commented:
> scuba testing

Thank you for your code suggestion and feedback. Please let me know if you need any assistance from my side 🙏

pranavo72bex commented:
> The scuba changes look a little inconsistent, which is unexpected. The separator "drifts" around in different cases in ways you wouldn't expect. I talked to @QuncCccccc about this and she did a deep dive and found the problem was with the separator not picking up the right style between dial and input mode. Which existed before this PR, but this PR caused the issue to actually show up.
> 
> @pranavo72bex So at no fault of this PR, I think it shouldn't be merged without that fix as well, to lessen the amount of scuba breakages between two separate PRs. I'll be willing to open a PR against this one with the needed fix, so they can all go together, but I likely won't be able to get to it for another week. Apologies.

You're welcome! I'm glad the insights were helpful. Let me know if there's anything else you'd like me to clarify or assist with. Always happy to collaborate! 😊

Piinks commented:
Caught up with @MitchellGoodwin offline, we want to take a deeper look at this to unblock its progress. Marking it as a draft for now while Mitchel is looking into it.

pranavo72bex commented:
> Caught up with @MitchellGoodwin offline, we want to take a deeper look at this to unblock its progress. Marking it as a draft for now while Mitchel is looking into it.

Thank you @Piinks . Please let me know if you need any assistance from my side

pranavo72bex commented:
Hello @MitchellGoodwin any update on this PR 💭

dkwingsmt commented:
@Piinks @MitchellGoodwin Have you taken a look at https://github.com/flutter/flutter/pull/157402#issuecomment-2578332790 ?

MitchellGoodwin commented:
I was looking into this, and was having trouble getting the g3 failures to be consistent when reproducing. I have not been able to put enough time into it, however.

Piinks commented:
Rebased for a fresh result from Google testing, cc @MitchellGoodwin 
@pranavo72bex heads up, it also looked like the analyzer was failing, let's see if it does again on this fresh run.

pranavo72bex commented:
Thank you to everyone who engaged with this pull request over the past several months. I initially opened this PR with the hope of contributing to the improvement of the Flutter TimePicker, specifically addressing the vertical alignment of the separator (":"), and followed through on feedback regarding golden tests and formatting.

However, after six months without resolution or further progress—and despite multiple iterations, rebases, and ongoing follow-ups—I’ve decided to close this PR and revert the associated changes on my end. I genuinely appreciate the early feedback and engagement, but I must admit I’m quite disappointed by the prolonged delays and lack of final decision-making.

I value this community and remain hopeful that future contributions will be met with clearer collaboration and resolution timelines.

Thank you again.

— @pranavo72bex

MitchellGoodwin commented:
Apologies @pranavo72bex, I'll take full responsibility on this. This change was unusually tricky to diagnose the behavior in Google testing compared to the relatively straightforward change this PR was making. Combined with my other work this was difficult to prioritize, but I should have gave it more time. And at bare minimum I should have communicated better. My apologies.

pranavo72bex commented:
> Apologies @pranavo72bex, I'll take full responsibility on this. This change was unusually tricky to diagnose the behavior in Google testing compared to the relatively straightforward change this PR was making. Combined with my other work this was difficult to prioritize, but I should have gave it more time. And at bare minimum I should have communicated better. My apologies.

Thank you for your thoughtful response, @MitchellGoodwin — I really appreciate your honesty and the context.

To help move things forward more efficiently, would it be possible for me to submit a separate PR that only includes the core fix (ensuring the separator is vertically centered) and leaves out the golden tests for now? I can then follow up with a second PR that focuses solely on adding the golden tests, which we can iterate on independently without blocking the main UI improvement.

Please let me know if that approach would be acceptable.

Thanks again for your support!

MitchellGoodwin commented:
> To help move things forward more efficiently, would it be possible for me to submit a separate PR that only includes the core fix (ensuring the separator is vertically centered) and leaves out the golden tests for now? I can then follow up with a second PR that focuses solely on adding the golden tests, which we can iterate on independently without blocking the main UI improvement.

The Google testing failures were because of golden tests that already existed within Google. The newly added one that was a part of this PR was not a problem. You are still more than welcome to reopen this PR.

pranavo72bex commented:
> > To help move things forward more efficiently, would it be possible for me to submit a separate PR that only includes the core fix (ensuring the separator is vertically centered) and leaves out the golden tests for now? I can then follow up with a second PR that focuses solely on adding the golden tests, which we can iterate on independently without blocking the main UI improvement.
> 
> The Google testing failures were because of golden tests that already existed within Google. The newly added one that was a part of this PR was not a problem. You are still more than welcome to reopen this PR.

I have reopen the PR https://github.com/flutter/flutter/pull/168441
